### PR TITLE
Argparse Module

### PR DESCRIPTION
**Pull request by briandowns**
### Well detailed description of the change :

Adds a module to the stdlib for parsing args.

Resolves: #570 

#

### Type of change:

- [ ] Bug fix

- [x] New feature

- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#

### Housekeeping

- [ ] Tests have been updated to reflect the changes done within this PR (if applicable).

- [ ] Documentation has been updated to reflect the changes done within this PR (if applicable).

#

### CODE REVIEW

**Comment by Jason2605** [comment]
Sorry been a busy few weeks, will have time to look at this later in the week!

**Comment by briandowns** [comment]
No worries. Thanks!

**Comment by Jason2605** [comment]
Apologies on the delay on this!!! I’m away this weekend but I will try my best to give it a look Sunday, sorry!

File: src/optionals/argparse/argparse.du [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    var required = [];
+
+    init(var name, var desc, var userUsage) {
+        this.args = Args(name, desc);
+    }
+
+    private flagExists(flag) {
+        for (var i = 0; i < this.preArgs.len(); i+=1) {
+            if (this.preArgs[i]['flag'] == flag) {
+                return true;\
```
Comment (Jason2605) [inline src/optionals/argparse/argparse.du:24]: I see there's a couple of backslashes in here! This is my fault for the script not being documented at all but there is actually a [generate.du](https://github.com/dictu-lang/Dictu/blob/develop/scripts/generate.du) file that'll deal with creating these header files for you!

File: src/optionals/argparse/argparse-source.h [hunk 2]
```diff
@@ -0,0 +1,133 @@
+#define DICTU_ENV_SOURCE "import Argparse; \
```
Comment (Jason2605) [inline src/optionals/argparse/argparse-source.h:1]: We will want `DICTU_ARGPARSE_SOURCE` here but using the generate script (once we add the argparse path should resolve this)

File: src/optionals/argparse/argparse.c [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#include "argparse.h"
+
+#include "argparse-source.h"
+
+Value createArgParseModule(DictuVM *vm) {
+    ObjClosure *closure = compileModuleToClosure(vm, "Argparse", DICTU_ENV_SOURCE);
```
Comment (Jason2605) [inline src/optionals/argparse/argparse.c:6]:
```suggestion
    ObjClosure *closure = compileModuleToClosure(vm, "Argparse", DICTU_ARGPARSE_SOURCE);
```

**Comment by briandowns** [comment]
Thanks for the review. Looks like everything is sorted.

File: src/optionals/argparse/argparse.du [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    private addFlag(flagType, flag, desc, required, ...metavar) {
+        if (not this.flagExists(flag)) {
+            this.preArgs.push({
+                'type': flagType,
+                'flag': flag,
+                'desc': desc,
+                'required': required,
+                'metavr': metavar
```
Comment (Jason2605) [inline src/optionals/argparse/argparse.du:38]: Is this a typo?

File: src/optionals/argparse/argparse.du [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+    parse() {
+        for (var i = 0; i < System.argv.len(); i+=1) {
+            for (var j = 0; j < this.preArgs.len(); j+=1) {
+                if (System.argv[i] == this.preArgs[j]['flag']) {
+                    if (this.preArgs[j]['type'] == 'bool') {
+                        this.args.setAttribute(this.preArgs[j]['flag'].replace('-', ''), true);
+                    } else if (this.preArgs[j]['type'] == 'list') {
+                        if (i == (System.argv.len() - 1) or System.argv[i+1][0] == '-') {
+                            return Error('{} requires an argument');
```
Comment (Jason2605) [inline src/optionals/argparse/argparse.du:101]: Missing a `.format()` here

File: src/optionals/argparse/argparse.du [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+                        }
+
+                        this.args.setAttribute(this.preArgs[j]['flag'].replace('-', ''), System.argv[i+1].split(','));
+                    } else if (this.preArgs[j]['type'] == 'number') {
+                        if (i == (System.argv.len() - 1) or System.argv[i+1][0] == '-') {
+                            return Error('{} requires an argument');
```
Comment (Jason2605) [inline src/optionals/argparse/argparse.du:107]: Missing a `.format()` here

File: src/optionals/argparse/argparse.du [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+                        }
+                        const res = System.argv[i+1].toNumber();
+                        if (not res.success()) {
+                            return Error('{} arg must be a number'.format(System.argv[i]));
+                        }
+
+                        this.args.setAttribute(this.preArgs[j]['flag'].replace('-', ''), res.unwrap());
+                    } else {
+                        if (i == (System.argv.len() - 1) or System.argv[i+1][0] == '-') {
+                            return Error('{} requires an argument');
```
Comment (Jason2605) [inline src/optionals/argparse/argparse.du:117]: Missing a `.format()` here

**Review by Jason2605 — COMMENTED** [review]
Couple of comments on here!

Be good in the future too on the long specifiers if we can support the = syntax, e.g `--port=100` rather than `--port 100` too, but that's just a "nice to have"

**Comment by Jason2605** [comment]
Thanks for this!

**Pull request merged**